QModManager/API/ModLoading/QModPatchAttributeBase.cs

```diff
@@ -0,0 +1,99 @@
+﻿namespace QModManager.API
+{
+    using System;
+    using System.Reflection;
+    using System.Security.Cryptography;
+    using System.Text;
+    using QModManager.Patching;
+
+    /// <summary>
+    /// Base class to all attributes that identify QMod patch methods.
+    /// </summary>
+    /// <seealso cref="Attribute" />
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = false, Inherited = false)]
+    public abstract class QModPatchAttributeBase : Attribute
+    {
+        internal PatchingOrder PatchOrder { get; }
+
+        private readonly string _secretPasword;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="QModPatchAttributeBase"/> class.
+        /// </summary>
+        /// <param name="patchOrder">The patch order.</param>
+        internal QModPatchAttributeBase(PatchingOrder patchOrder)
+        {
+            this.PatchOrder = patchOrder;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="QModPatchAttributeBase" /> class.
+        /// </summary>
+        /// <param name="patchOrder">The patch order.</param>
+        /// <param name="secretPasword">The secret pasword.</param>
+        /// <exception cref="FatalPatchingException">This modder has not read the documentation and should not be using prepatch/postpatch functions.</exception>
+        internal QModPatchAttributeBase(PatchingOrder patchOrder, string secretPasword) : this(patchOrder)
+        {
+            if (string.IsNullOrEmpty(secretPasword))
+                throw new FatalPatchingException("This modder has not read the documentation and should not be using prepatch/postpatch functions.");
+
+            _secretPasword = secretPasword;
+        }
+
+        /// <summary>
+        /// Validates the that modder has read the documentation.
+        /// </summary>
+        /// <param name="method">The method.</param>
+        /// <exception cref="FatalPatchingException">This modder has not read the documentation and should not be using prepatch/postpatch functions.</exception>
+        internal void ValidateThatModderHasReadTheDocumentation(MethodInfo method)
```

Comment:
**TODO** - Update this to _also_ use the QMod Id, as it may be different from the Assembly name and is guaranteed to be unique within the ordering while the Assembly name may not be.